formats/cbor/commonMain/src/kotlinx/serialization/cbor/ByteString.kt

```diff
@@ -0,0 +1,34 @@
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.Decoder
+import kotlinx.serialization.Encoder
+import kotlinx.serialization.SerialDescriptor
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.StructureKind
+
+public class ByteString(
+    public val bytes: ByteArray
+) {
+    public val size: Int = bytes.size
+
+    @Serializer(forClass = ByteString::class)
+    public companion object : CborKSerializer<ByteString> {
+        override val descriptor: SerialDescriptor = SerialDescriptor(
+            serialName = "kotlinx.serialization.ByteString",
+            kind = StructureKind.BYTE_STRING
+        )
+
+        override fun serialize(encoder: CborEncoder, value: ByteString): Unit = encoder.encodeByteString(value)
+        override fun deserialize(decoder: CborDecoder): ByteString = decoder.decodeByteString()
+
+        override fun serialize(encoder: Encoder, value: ByteString) {
+            check(encoder is CborEncoder) { "CBOR encoding must use a CborEncoder" }
+            serialize(encoder, value)
+        }
```

Comment:
Could this just read?:

```suggestion
        override fun serialize(encoder: Encoder, value: ByteString) {
            throw UnsupportedOperation("CBOR encoding must use a CborEncoder")
        }
```

_(or similar exception)_

With the overloaded function, a `CborDecoder` would never land in this function, so makes sense to just throw right away rather than doing the check.

Reply:
I don't think we're quite overloading, right? I guess in theory, but not technically since the first parameter is a different type (`CborEncoder` vs `Encoder`). Or does that still qualify as overloading? Anyway, overloading was what I had hoped to do, but couldn't achieve with this approach (because I'm probably missing something, but I don't know what).

The method in `Cbor.kt` that starts the serialization process does create an instance of `CborEncoder`/`CborDecoder`, but those references are passed to functions that only deal with `Encoder`/`Decoder` types. So `serialize(encoder: Encoder, value: ByteString)` does get called since that's the only way the non-format specific interfaces and class know to how invoke the serializer methods.